### PR TITLE
getClosestPlotData() should exclude off-plot data

### PR DESCRIPTION
**Pull request by istarnx**
For consideration in the getClosestPlotData() algorithm, data should be currently visible on the plot. This PR adds the protected _isVisibleOnPlot() method so that each type can implement its own notion of visibility.

### CODE REVIEW

File: src/components/plots/barPlot.ts [hunk 1]
```diff
@@ -179,7 +178,7 @@ export module Plot {
           var secondaryDist = 0;
 
           // if we're inside a bar, distance in both directions should stay 0
-          if (!_Util.Methods.intersectsBBox(queryPoint.x, queryPoint.y, barBBox)) {
+          if (!_Util.Methods.intersectsBBox(queryPoint.x, queryPoint.y, bar.getBBox())) {
```
Comment (bluong) [inline src/components/plots/barPlot.ts]: barBBox here please

File: src/components/plots/barPlot.ts [hunk 2]
```diff
@@ -214,6 +213,14 @@ export module Plot {
       };
     }
 
+    protected _isVisibleOnPlot(datum: any, pixelPoint: Point, selection: D3.Selection): boolean {
+      var chartXExtent = { min: 0, max: this.width() };
+      var chartYExtent = { min: 0, max: this.height() };
```
Comment (bluong) [inline src/components/plots/barPlot.ts]: Use plotXExtent

File: test/components/plots/plotTests.ts [hunk 3]
```diff
@@ -250,54 +250,6 @@ describe("Plots", () => {
       svg.remove();
     });
 
-    it("getClosestPlotData", () => {
```
Comment (jtlan) [inline test/components/plots/plotTests.ts]: We'll file a testing issue on this.

File: src/components/plots/linePlot.ts [hunk 4]
```diff
@@ -193,7 +196,7 @@ export module Plot {
 
           if (xDist === minXDist && yDist === minYDist) {
             closestData.push(plotData.data[index]);
```
Comment (bluong) [inline src/components/plots/linePlot.ts]: If we've defined a datum variable above, let's push the datum in here.

File: src/components/plots/linePlot.ts [hunk 5]
```diff
@@ -193,7 +196,7 @@ export module Plot {
...
-            closestPixelPoints.push(pxPt);
+            closestPixelPoints.push(pixelPoint);
             closestElements.push(plotData.selection[0][0]);
```
Comment (bluong) [inline src/components/plots/linePlot.ts]: Same as above

File: src/components/plots/scatterPlot.ts [hunk 6]
```diff
@@ -115,6 +115,22 @@ export module Plot {
       };
     }
 
+    protected _isVisibleOnPlot(datum: any, pixelPoint: Point, selection: D3.Selection): boolean {
+      var chartXExtent = { min: 0, max: this.width() };
+      var chartYExtent = { min: 0, max: this.height() };
```
Comment (bluong) [inline src/components/plots/scatterPlot.ts]: Rename to plotXExtent

File: src/components/plots/barPlot.ts (hunk 2 above)
Comment (jtlan) [inline src/components/plots/barPlot.ts]: "xRange" and "yRange"? We use "extent" elsewhere to refer to a pair of min-max values in data-space.

**Pull request merged**